Sprint 15

ZakharVP opened the pull request.

Commit:
Точка остановки

Commit:
Выполнил задачу по 15 спринту

NaR0RoW commented on Tracker.xcodeproj/xcuserdata/zaharpancenko.xcuserdatad/xcdebugger/Breakpoints_v2.xcbkptlist:
Папку `xcuserdata` стоит игнорировать, кладя в `.gitignore`. В ней хранится информация о состоянии проекта (какой файл открыт, какой таргет выбран и т.п.), брейкпоинты и схемы для текущего пользователя `Xcode`.
Про `gitignore` можно прочитать вот [тут](https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files)

NaR0RoW commented on Tracker/AppDelegate.swift:
```diff
@@ -16,6 +16,20 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
         // Override point for customization after application launch.
         
+        lazy var persistentContainer: NSPersistentContainer = {
```
Стоит удалить этот `persistentContainer` и еще один внутри `AppDelegate`, т.к. у нас уже есть `CoreDataManager`, стоит использовать его

NaR0RoW commented on Tracker/Controllers/10_ChoiceCategoryViewController.swift:
```diff
@@ -10,11 +10,12 @@ import UIKit
 class ChoiceCategoryViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {
```
Стоит добавить модификатор `final`, чтобы вызовы методов работали быстрее.
Но это уже со звездочкой, хотя на собеседованиях спрашивают. Можно почитать про [диспачеризацию тут](https://habr.com/ru/articles/673400/)

NaR0RoW commented on Tracker/Controllers/10_ChoiceCategoryViewController.swift:
```diff
@@ -142,8 +141,13 @@ class ChoiceCategoryViewController: UIViewController, UITableViewDelegate, UITab
         return min(calculatedHeight, maxHeight)
     }
     
+    private func loadCategories() {
+           categories = categoryStore.fetchAllCategories()
+           setupUpdateUI()
+       }
```
Форматирование кода немного поехало, можно выделить это место и воспользоваться автоматическим форматированием в Xcode `control + I`

NaR0RoW commented on Tracker/Controllers/27_IrregularEventTrackerViewController.swift:
```diff
@@ -446,8 +442,15 @@ class IrregularEventTrackerViewController: UIViewController, UICollectionViewDel
     
     func updateButtonCreateState() {
-        let isEnabled = !trackerTitle.isEmpty && selectedCategory != nil
-        buttonCreate.backgroundColor = isEnabled ? .black : .gray
+        //let isEnabled = !trackerTitle.isEmpty && selectedCategory != nil
+        //buttonCreate.backgroundColor = isEnabled ? .black : .gray
+        
+        let isEnabled = !trackerTitle.isEmpty &&
+                           selectedCategory != nil &&
+                           (selectedEmoji != nil) &&
+                           (selectedColor != nil)
+            buttonCreate.backgroundColor = isEnabled ? .black : .gray
+            buttonCreate.isEnabled = isEnabled
```
Тот тоже форматирование кода поехало

NaR0RoW commented on Tracker/Controllers/27_IrregularEventTrackerViewController.swift:
```diff
@@ -446,8 +442,15 @@ class IrregularEventTrackerViewController: UIViewController, UICollectionViewDel
     func updateButtonCreateState() {
-        let isEnabled = !trackerTitle.isEmpty && selectedCategory != nil
-        buttonCreate.backgroundColor = isEnabled ? .black : .gray
+        //let isEnabled = !trackerTitle.isEmpty && selectedCategory != nil
```
Лучше удалять закомментированный код перед review

NaR0RoW commented on Tracker/Controllers/ViewController.swift:
```diff
@@ -8,6 +8,11 @@
 import UIKit
 
 class ViewController: UIViewController {
```
Стоит отметить как `final`. Перепроверь пожалуйста остальные классы в коде. Если класс не наследуется, то лучше отметить его как `final` для ускорения работы компилятора

NaR0RoW replied:
Стоит более осмысленные названия использовать. Из названия не понятно, за что отвечает данный view controller

NaR0RoW commented on Tracker/Controllers/ViewController.swift:
```diff
@@ -123,6 +129,16 @@ class ViewController: UIViewController {
         navigationController?.setNavigationBarHidden(false, animated: animated)
     }
     
+    private func loadData() {
+            // Загружаем категории и трекеры
+            categories = categoryStore.fetchAllCategories()
+            
+            // Загружаем выполненные трекеры для выбранной даты
+            loadCompletedTrackers(for: datePicker.date)
+            
+            updateUI()
+        }
```
Давай везде поправим выравнивание кода

NaR0RoW commented on Tracker/Store/CoreDataManager.swift:
```diff
@@ -0,0 +1,36 @@
+        let container = NSPersistentContainer(name: "Tracker")
+        container.loadPersistentStores { _, error in
+            if let error = error {
+                fatalError("Не могу загрузить persistent stores: \(error)")
```
По своему опыту хотел бы отметить, что в продакшене не используют `fatalError`, советую заменить на `assertionFailure`, [тут подробнее](https://www.hackingwithswift.com/plus/intermediate-swift/understanding-assertions)

NaR0RoW commented on Tracker/Store/TrackerStore.swift:
```diff
@@ -0,0 +1,86 @@
+        trackerCoreData.dateCreateTracker = Date()
+        trackerCoreData.category = category
+        
+        if !tracker.shedule.isEmpty {
```
Ошибка в правописании:
`shedule` -> `schedule`

Лучше включить [настройку](https://sarunw.com/posts/spell-checking-in-xcode/), чтобы `Xcode` подсвечивал ошибки в правописании

Commit:
Удалил xcuserdata из кеша git

Commit:
Изменил получение обновлений от Core Data на NSFetchedResultsController

Commit:
refactor (coredata): Избыточное создание persistentContainer

Удалил избыточное создание persistentContainer

Commit:
refactor(classes): Добавил final классам

Добавил final классам которые отвечают за показ UI экранов и не имеют подчиненных класов

Commit:
style(format): auto-format строк

Сделал auto-format  строк при помощи комбинации ctrl+I по всему проекту.

Commit:
refactor(logging): изменил fatalError на print

Изменил вывод сообщений fatalError() на print("")

Commit:
chore(git): ignore Xcode user data

Очистил кэш для отслеживания изменений по файлам.
